tesseract_core/runtime/array_encoding.py

```diff
@@ -381,7 +384,15 @@ def decode_array(
 
         # keep checking for "raw" for backwards compat
         elif val.data.encoding in {"json", "raw"}:
-            data = np.asarray(val.data.buffer, dtype=None).reshape(val.shape)
+            data = np.asarray(val.data.buffer).reshape(val.shape)
+            if np.issubdtype(data.dtype, np.floating) and np.issubdtype(
+                val.dtype, np.integer
+            ):
+                if np.any(data % 1):
+                    raise ValueError(
+                        f"Expected integer data, but got floating point data: {data}"
+                    )
+            data = data.astype(val.dtype, casting="unsafe")
```

Comment:
What's wrong with the following?
```suggestion
data = data.astype(val.dtype, casting="same_kind")
```
I'm sure you have some edge cases in mind, but not sure what these would be as `same_kind` does allow conversion from `np.integer` to `np.floating` (It's a shame that `asarray` doesn't allow a casting `kwarg`.)

> ‘same_kind’ means only safe casts or casts within a kind, like float64 to float32, are allowed.

Reply:
We're sometimes converting floats to integers here.